### PR TITLE
Ability to communicate environment variables from leader to worker (resolves #441, resolves #421)

### LINKED ISSUE #421: Incorrect catch for missing .toilAzureConfig
The Azure job store tries to give a useful message when it can't read your .toilAzureConfig for the key for the Azure job store storage account. However, it is looking for a `NoOptionError`, and if the entire file is missing instead a `NoSectionError` is raised, producing an error like this:

```


DEBUG:toil.batchSystems.mesos.executor:Starting executor
I0924 22:47:08.587648 43191 exec.cpp:132] Version: 0.22.0
I0924 22:47:08.592193 43192 exec.cpp:206] Executor registered on slave 20150924-214557-83886090-5050-44087-S8
INFO:toil.batchSystems.mesos.executor:Registered with framework
DEBUG:toil.batchSystems.mesos.executor:Running task 0
DEBUG:toil.batchSystems.mesos.executor:Sending status update ...
DEBUG:toil.batchSystems.mesos.executor:Sent status update
DEBUG:toil.batchSystems.mesos.executor:Invoking command: '/usr/bin/python2.7 -E /usr/local/lib/python2.7/dist-packages/toil/worker.py azure:hgvm:tree2 683b1010_55c3_4978_a7e7_853f17ed1803'
DEBUG:toil.batchSystems.mesos.executor:Sent stats message
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/toil/worker.py", line 408, in <module>
    main()
  File "/usr/local/lib/python2.7/dist-packages/toil/worker.py", line 89, in main
    jobStore = loadJobStore(jobStoreString)
  File "/usr/local/lib/python2.7/dist-packages/toil/common.py", line 366, in loadJobStore
    return AzureJobStore( account, namePrefix, config=config )
  File "/usr/local/lib/python2.7/dist-packages/toil/jobStores/azureJobStore.py", line 78, in __init__
    account_key = _fetchAzureAccountKey(accountName)
  File "/usr/local/lib/python2.7/dist-packages/toil/jobStores/azureJobStore.py", line 63, in _fetchAzureAccountKey
    return configParser.get('AzureStorageCredentials', accountName)
  File "/usr/lib/python2.7/ConfigParser.py", line 330, in get
    raise NoSectionError(section)
ConfigParser.NoSectionError: No section: 'AzureStorageCredentials'
DEBUG:toil.batchSystems.mesos.executor:Sending status update ...
DEBUG:toil.batchSystems.mesos.executor:Sent status update
I0924 22:47:09.738494 43200 exec.cpp:379] Executor asked to shutdown
CRITICAL:toil.batchSystems.mesos.executor:Shutting down executor...
CRITICAL:toil.batchSystems.mesos.executor:Executor shut down
```

I'm observing this on my Mesos slave nodes, which don't share a home directory with my master.

Additionally, although I have set `AZURE_ACCOUNT_KEY` on the master, it doesn't seem to be propagating to the slaves to prevent this (perhaps because the environment loading and saving happens in the worker, using stuff stored in the job store?).

### LINKED ISSUE #441: Implement AbstractBatchSystem.setEnv
For details see 

https://github.com/BD2KGenomics/toil/pull/437#issuecomment-144546077

and 

https://github.com/BD2KGenomics/toil/issues/421#issuecomment-143406075

### PR DESCRIPTION
**Pull request by JakeNarkizian**
Implemented setEnv in mesos, and single machine batch systems. Added command line option for setEnv. Added getEnv in abstractJobStore and called it from leader.py to iterate and set variables in batch system,

Resolves #441 and #421

### CODE REVIEW

File: src/toil/jobStores/abstractJobStore.py [hunk 1]
```diff
@@ -223,6 +223,9 @@ def exists(self, jobStoreID):
     # One year should be sufficient to finish any pipeline ;-)
     publicUrlExpiration = timedelta(days=365)
 
+    def getEnv(self):
+        return ['AZURE_ACCOUNT_KEY'] #not sure what other env variables to ask for
```
Comment (hannes-ucsc) [inline src/toil/jobStores/abstractJobStore.py]: This is an Azure-specific detail that has no place in the abstract job store. Override this method in the Azure job store and change this implementation to return an empty collection.

File: src/toil/batchSystems/abstractBatchSystem.py [hunk 2]
```diff
@@ -99,6 +99,12 @@ def shutdown(self):
         """
         raise NotImplementedError('Abstract Method: shutdown')
 
+    def setEnv(self,name, value=None):
+        """Sets environment variables. If no value is provided it will be looked up
```
Comment (hannes-ucsc) [inline src/toil/batchSystems/abstractBatchSystem.py]: Newline after """

Comment (hannes-ucsc) [inline src/toil/batchSystems/abstractBatchSystem.py]: Needs more substance. Describe what the effect of this method on the running jobs is.

File: src/toil/batchSystems/abstractBatchSystem.py [hunk 3]
```diff
@@ -99,6 +99,12 @@ def shutdown(self):
...
+        from the current environment.
+        """
+        raise NotImplementedError('Abstract method: setEnv')
```
Comment (hannes-ucsc) [inline src/toil/batchSystems/abstractBatchSystem.py]: No need for the message argument to the exception constructor. It will be clear from the trace back which method was left unimplemented. Also, the ABC mechanism will catch unimplemented methods much earlier, at instantiation time.

File: src/toil/batchSystems/mesos/batchSystem.py [hunk 4]
```diff
@@ -77,6 +77,9 @@ def __init__(self, config, maxCores, maxMemory, maxDisk, masterIP,
         # Dict of launched jobIDs to TaskData named tuple. Contains start time, executorID, and slaveID.
         self.runningJobMap = {}
 
+        # Dict of envirnment variables
```
Comment (hannes-ucsc) [inline src/toil/batchSystems/mesos/batchSystem.py]: Spelling. And lets change "dict" to "dictionary"

File: src/toil/batchSystems/mesos/batchSystem.py [hunk 5]
```diff
@@ -262,6 +265,15 @@ def shutdown(self):
         if driver_result != mesos_pb2.DRIVER_STOPPED:
             raise RuntimeError("Mesos driver failed with %i", driver_result)
 
+    def setEnv(self,name, value=None):
```
Comment (hannes-ucsc) [inline src/toil/batchSystems/mesos/batchSystem.py]: space after comma

File: src/toil/batchSystems/mesos/batchSystem.py [hunk 6]
```diff
@@ -262,6 +265,15 @@ def shutdown(self):
...
+        """
+        Sets environment variables into dictionary for each job.
```
Comment (hannes-ucsc) [inline src/toil/batchSystems/mesos/batchSystem.py]: Only need to provide a docstring if the docstring in the overridden method needs to be amended, e.g. if you are modifying the contract defined in that method signature/docstring.

File: src/toil/batchSystems/mesos/executor.py [hunk 7]
```diff
@@ -124,7 +124,7 @@ def runJob(job):
                 job.userScript.register()
             log.debug("Invoking command: '%s'", job.command)
             with self.popenLock:
-                return subprocess.Popen(job.command, shell=True)
+                return subprocess.Popen(job.command, shell=True, env=job.environment)
```
Comment (hannes-ucsc) [inline src/toil/batchSystems/mesos/executor.py]: Here you need to mix the current environment with job.environment. Otherwise, the current environment would be lost and the only variables requested via setEnv would be present. But we want the current environment variables *plus* the ones requested via setEnv.

File: src/toil/batchSystems/singleMachine.py [hunk 8]
```diff
@@ -193,6 +195,14 @@ def shutdown(self):
         for thread in self.workerThreads:
             thread.join()
 
+    def setEnv(self,name, value=None):
+        """
+        Sets environment variables into dictionary for each job.
```
Comment (hannes-ucsc) [inline src/toil/batchSystems/singleMachine.py]: Same here: no docstring necessary.

File: src/toil/batchSystems/mesos/batchSystem.py [hunk 9]
```diff
@@ -262,6 +265,15 @@ def shutdown(self):
...
+        """
+        if value is None and name not in os.environ:
+            raise RuntimeError("%s does not exist in current environment", name)
+
+        self.environment[name] = value or os.environ[name]
```
Comment (hannes-ucsc) [inline src/toil/batchSystems/mesos/batchSystem.py]: The way this method is implemented it does two lookups into the os.environ dictionary. This can introduce race conditions and is not optimal. It also checks `value` twice, but differently, causing this method to fail in certain cases if the value is "". Rewrite this method to only use one lookup and one conditional on value. Hint: use a try-except statement. The last line of the method will read `self.environment[name] = value`

File: src/toil/common.py [hunk 10]
```diff
@@ -213,6 +215,7 @@ def _addOptions(addGroupFn, config):
     addOptionFn("--maxParasolBatches", dest="maxParasolBatches", default=None,
                 help="Maximum number of batches Parasol is allowed to create - a batch \
                 is created for each job that has a unique set of resource requirements. Default=%i" % config.maxParasolBatches)
+    addOptionFn("--setEnv", dest="setEnv", default={}, help="Set of arbitrary environment variables for workers. Default=%i" % config.setEnv)
```
Comment (hannes-ucsc) [inline src/toil/common.py]: Did you actually try this? How do I specify multiple variables on the same script invocation?

File: src/toil/leader.py [hunk 11]
```diff
@@ -335,6 +336,10 @@ def mainLoop(config, batchSystem, jobStore, rootJobWrapper):
     logger.info("Found %s jobs to start and %i jobs with successors to run",
                 len(toilState.updatedJobs), len(toilState.successorCounts))
 
+    #Set proper environment variables
+    for v in getEnv():
```
Comment (hannes-ucsc) [inline src/toil/leader.py]: getEnv() should be a method of the batch system class. It can't be a statically defined function in the module. Look at the other methods in AbstractBatchSystem and model your new method after those.

**Comment by hannes-ucsc** [comment]
Aside from the line nodes, this functionality needs a unit test. The test should prove for every supported batch system that a variable injected via setEnv is indeed present when a job runs.

File: src/toil/jobStores/abstractJobStore.py (hunk 1 above)
Comment (hannes-ucsc) [inline src/toil/jobStores/abstractJobStore.py]: The return value must be a dictionary, since you need to support entries with fixed value that is not looked up from the environment. Most importantly, the AzureJobStore implementation of this method needs to return a value for AZURE_ACCOUNT_KEY since that value is read from a config file, not the current directory. I outlined this in my original email.

**Comment by hannes-ucsc** [comment]
The tests failed with an obvious import error. This indicates to me that you never ran the the unit tests on your box before committing. But you should. At least the basic ones.

File: src/toil/leader.py (hunk 11 above)
Comment (JakeNarkizian) [inline src/toil/leader.py]: I'm slightly confused, should getEnv() be a function in both the batch system and jobstore? if so what is the purpose of AbstractBatchSystem.getEnv()?

Comment (hannes-ucsc) [inline src/toil/leader.py]: Yep, I mixed up batch system and job store. Here's what the line not should have read:

> getEnv() should be a method of the job store class. It can't be a statically defined function in a module. Look at the other methods in AbstractJobStore and model your new method after those.

**Comment by hannes-ucsc** [comment]
I won't be able to review this until the tests pass.

The commit history is a mess: the commits from the master branch (e.g. 8c646c6) should not appear in this branch and the "reverted changes that were accidentally committed" commit should not be there. You should probably squash all commits and rebase onto the latest master (as you always should with single-author PRs).

File: src/toil/batchSystems/singleMachine.py [hunk 12]
```diff
@@ -85,6 +85,8 @@ def __init__(self, config, maxCores, maxMemory, maxDisk):
         self.coreFractions = ResourcePool(self.numWorkers)
         # A lock to work around the lack of thread-safety in Python's subprocess module
         self.popenLock = Lock()
+        # A dictionary of environment variables to be passed to the job subprocess
+        self.evironment = {}
```
Comment (hannes-ucsc) [inline src/toil/batchSystems/singleMachine.py]: Spelling

File: src/toil/batchSystems/singleMachine.py [hunk 13]
```diff
@@ -193,6 +195,16 @@ def shutdown(self):
         for thread in self.workerThreads:
             thread.join()
 
+    def setEnv(self, name, value=None):
+        try:
+            if name is "":
```
Comment (hannes-ucsc) [inline src/toil/batchSystems/singleMachine.py]: Why is this necessary?

Also, you must use == for strings.

File: src/toil/batchSystems/singleMachine.py [hunk 14]
```diff
@@ -193,6 +195,16 @@ def shutdown(self):
...
+                return
+            if value is None and name is not "":
```
Comment (hannes-ucsc) [inline src/toil/batchSystems/singleMachine.py]: 2nd term is a dupe.

File: src/toil/batchSystems/singleMachine.py (hunk 13 above)
Comment (hannes-ucsc) [inline src/toil/batchSystems/singleMachine.py]: Also, `if` and `try` should be swapped.

File: src/toil/common.py [hunk 15]
```diff
@@ -132,6 +133,24 @@ def f(x):
         setOption("masterIP") 
         setOption("parasolCommand")
         setOption("maxParasolBatches", int, iC(1))
+
+        def splitEnvVarsToDict(envVars):
+            """
+            Parses of list of environment variables and values. Names and their values are to be
+            separated by a space, an entirely different variable is to be separated by a comma.
+            e.g. --setEnv var1 val1, var2, var3 val3
```
Comment (hannes-ucsc) [inline src/toil/common.py]: 1) This is an unconventional syntax. For variables, the convention is `name=value`
2) This syntax does not support spaces in values, which are more common than you might think.
3) argparse has the ability to mention an option multiple times and it will collect all occurrences in a list (see https://docs.python.org/2.7/library/argparse.html#action action='append')

Comment (hannes-ucsc) [inline src/toil/common.py]: If we were to go with this method (which I don't think we need to), I would want to see comprehensive doctests on it.

File: src/toil/jobStores/abstractJobStore.py [hunk 16]
```diff
@@ -114,6 +114,13 @@ def deleteJobStore(self):
         """
         raise NotImplementedError()
 
+    def getEnv(self):
+        """
+        Returns a list of relevant environment variables so that the leader can iterate over the list
```
Comment (hannes-ucsc) [inline src/toil/jobStores/abstractJobStore.py]: It returns a dict not a list

Comment (hannes-ucsc) [inline src/toil/jobStores/abstractJobStore.py]: And don't mention what a potential caller would want to do with the return value. That is their business. Take a step back and ask yourself "Why are we doing this?". The docstring should describe the *purpose* of the method. What is the intent behind it? Why would a subclass want to implement this method?

**Comment by hannes-ucsc** [comment]
Time is running out on this. We need to get the release out the door. I will have to take this over sometime tomorrow. If I do, don't consider your attempt a failure since I will be building on top of what you did.

**Comment by hannes-ucsc** [comment]
Tests still fail (there are some failures, and overall the build hangs), commits are not squashed and include commits from other branch.

**Comment by hannes-ucsc** [comment]
You should focus on finding out why `src/toil/test/mesos/mesosTest.py::MesosTest::test_hello_world` hangs. The hang prevents us from getting a proper test report from Jenkins. The test report will enable us to figure out why the other tests are failing.

**Comment by hannes-ucsc** [comment]
OK, I will take it from here.
